### ISSUE TITLE
Support git deduplication for new repos

### ISSUE THREAD

**Issue by indradhanush** [sourcegraph/sourcegraph#54741]
Part of #40448.

- [ ] **Task 1** Create (and merge :D) a PR and test potentially critical changes related to `addrForRepo` in production, this includes
    - [x] DB schema
    - [x] site config
    - [x] addrForRepo changes
    - [x] caching of addresses of deduplicated forks
    - [x] cache hit and miss metrics
    - [ ] **WIP** Observe in production (guarded behind the site config check)
- [ ] **Task 2** Main feature: Fix todos, hacks, add tests in #54663
    - [ ] Make this work for repos (and their forks) that would naturally clone on different shards
    - [ ] Fix progress bug in the UI (empty repository seen after the repo has been cloned)

Phase 1 of [RFC 798 Review: Dedupe Git Objects Across Forks](https://docs.google.com/document/d/18phDXepUP35FtsjYkzLRYXH19NU6yN-_IIixPgDPUAE/edit#heading=h.va7lesvrcble.).

## Follow ups to task 1

- [x] Evaluate query plan of current query added in #55001 ([see comment](https://github.com/sourcegraph/sourcegraph/pull/55001/files/f67282876761e7068597f2afb778382cb536c83c))
- [x] #55116 
- [x] Add code to conditionally bypass the cache 100% of the time and observe pressure created on the DB as a result and watch for regressions (https://github.com/sourcegraph/sourcegraph/pull/55188)
- [x] #55420
- [ ] #55422
- [x] #55421
- [ ] #55423

## Follow ups to task 2
- [ ] #55425
- [ ] #55426
- [x] #55427

## Refactoring tasks

- [ ] Update docs to AddrForRepo and include decision tree
- [ ] #55075 
- [ ] #55077 (only do this after the phase 2 is completed, at the moment an issue for phase 2 does not exist, migrate this to that issue when it's created and before this issue is closed)

**Comment by indradhanush** [comment]
**Git deduplication mid-week project update**

Yesterday, @keegancsmith and I found a blocker that needs to be resolved before we can make `AddrForRepo` lookup the `shard_id` in the DB - which is that the `gitserver.NewClient` needs to support accepting a DB so that it can then be propagated to the `GitserverAddresses` struct - and then `AddrForRepo` the method of this struct can use the DB handler. As a result we now have started a migration with mostly mechanical changes in this PR:

👉 https://github.com/sourcegraph/sourcegraph/pull/54788. (See Keegan’s [notes](https://github.com/sourcegraph/sourcegraph/pull/54788#issuecomment-1630948004) on the PR that describes the how we arrived eventually at that stage for more context.)

I had started separating the DB schema changes and the `AddrForRepo` related changes in this PR: https://github.com/sourcegraph/sourcegraph/pull/54774, but it remains blocked until we can complete the migration of all uses of `gitserver.NewClient` to use the DB handler. In the meantime I’ve been chugging away at making API improvements that Keegan and I discussed in the [original WIP PR](https://github.com/sourcegraph/sourcegraph/pull/54663/) (you can ignore this for the time being) that I will pull out once the migration is complete.


⚽  **Goals for the rest of the week**
- [ ] Complete migration of gitserver.NewClient
- [ ] Complete PR to separate out DB schema, site config changes and adding statefulness to AddrForRepo as planned above

- [ ] **Stretch goal:** Get the core deduplication code ready for review with fixes and improvements that have been discussed

Another great win that came out of our pairing session yesterday was that we most likely know why the cloning of forks into the same shard was not working 100% of the time and with the migration work and goal 2 described above we should be able to resolve it although I haven’t managed to test it yet because of the blocker described above.

**Comment by indradhanush** [comment]
Migration of gitserver.NewClient will be complete when the following PRs are reviewed and merged:
- [ ] #54850 
- [ ] #54867

**Comment by indradhanush** [comment]
**Update Monday July 17, 2023**

Last week I missed the goals set mid-week [above](https://github.com/sourcegraph/sourcegraph/issues/54741#issuecomment-1632189891) because the migration of `gitserver.NewClient` took a day longer than initially estimated. 

On Friday after the migration was complete I was able to make progress towards goal 2 - by adding statefulness to `AddrForRepo` [here](https://github.com/sourcegraph/sourcegraph/pull/54663/commits/bccb54a4d2b30cfef946b932a0f63d491b32907e) and I need half a day more to finish writing tests and create a separate PR by picking changes out from here before I can request reviews.

⚽ **Goals for the week**

- [ ] Complete PR that will make `AddrForRepo` stateful which includes the site config and DB schema changes to gitserver and related gitserver DB API methods
- [ ] Finish phase 1 of RFC wherein git repo deduplication works for new clones and is running on production

**Comment by indradhanush** [comment]
PR for task 1 (also goal 1 of this week) is here: https://github.com/sourcegraph/sourcegraph/pull/55001

**Comment by indradhanush** [comment]
### Update Monday July 24, 2023

On Friday, we deployed the latest changes on scaletesting, added an entry to `repositories.deduplicateForks` and set `experimentalFeatuers.useGitserverAddressLookupCache` to `false` to trigger the alternate gitserver address lookup path and to force it to read from the DB every single time. Over the last three days, here are our findings:

1. The query gets called the most number of times by an order of magnitude greater than the second place and even more than the UTILITY command that shows up in query insights
2. The query is fast and does not add an extra load to the DB


**Query** insights since the deployment and the forced DB lookup was enabled on Friday

<img width="1879" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/2682729/6f274fe3-b49a-4f09-a34f-099956ad2117">

**Last 7 days**

<img width="1906" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/2682729/e178be8a-757a-4844-bf25-d834c212125c">



**Next step**

We removed the `experimentalFeatuers.useGitserverAddressLookupCache` config earlier today and are now observing the query insights, we expect the number of DB reads to fall but not get eliminated because we hard expire the cache every 15 minutes.

⚽ **Goals for the week**

Last week we missed the second goal because we did not plan for the code review of #55001 and the follow up work it could create. This was a big change and indeed needed careful attention (thanks to everyone who took time out to review, re-review and re-review the PR!). As a result, most of last week was focused on completing the first goal instead.

This week the plan is to drive this issue to completion and with the following scope:

- Code for git deduplication is successfully reviewed and merged
- We have begun testing on at least one production like instance (scaletesting or S2)
- Git deduplication works when a repository is added to the config, and it has not been cloned to Sourcegraph yet

The foundation of this work is already in #54663 and we need to update it with the latest main branch, resolving conflicts along the way and updating and fixing code. I have broken this down to three high level tasks:

- [ ] Rebasing, updating existing code, fixing FIXMEs and TODOs, fixing temporar HACKS: 8 hours
- [ ] Adding tests: 8 hours
- [ ] Code reviews and follow ups: 8 hours

So overall, the estimate for this is now 3 days.

🏋🏽‍♂️ **Stretch goals**

- [ ] Update RFC, addressing existing comments
- [ ] Update docs with how deduplication works

**Comment by indradhanush** [comment]
## Update Monday July 31, 2023

Last week, I set out the main goal to have the code for deduplicated clone merged and deployed to sourcegraph.com so that we could have also started testing on scaletesting at least. However a few surprises came my way in terms of some life events which took away a chunk of my time. Additionally we had some other technical surprises too. Namely:

1. #55428: When I disabled the config to bypass the address lookup cache I saw no cache hits on the metrics on scaletesting, something that worked locally. However this was not an immediate blocker to make code related progress on the main goal of supporting deduplicated clones.

2. #55429: I got tunnel vision on the main goal of supporting deduplicated git clones, that it slipped my mind that I needed to also update the code path that does subsequent fetches. This was a planned for task under the phase 1 in the RFC but it slipped my attention as I was focused on the two large milestones of this issue. 

3. #54663: The PR that contains all the code for deduplicated git cloning had a whole bunch of existing tests that were failing - something that I did not anticipate / estimate for. Because of an internal implementation change in how the GitHub client creates the `Repository` resource, I had to fix all existing tests and combing through them took some unplanned time out of my already compromised schedule due to the unexpected life event.


### ⚽ Goals for this week

The project is beginning to create more work out of existing work in the form of follow ups and important requirements and items that need investigation in order to guarantee reliability in production. As a result while I've been adding follow up items above in the description it was quickly becoming cumbersome to manage and get a good overview of what needs to be done next. 

Consequently, I took a small step back to do some project planning and created a project board [here](https://github.com/orgs/sourcegraph/projects/366/views/1?filterQuery=-status%3ADone&visibleFields=%5B%22Title%22%2C%22Assignees%22%2C%22Status%22%2C52219424%2C%22Labels%22%5D). I resisted the temptation to go overboard with planning and only created a milestone for the first phase and issues that need to be worked upon to achieve it (including the follow up items that have been discovered).

- [ ] Get reviews on #54663 and address them as they show up or create follow up items that are not blockers: **1d**
- [ ] https://github.com/sourcegraph/sourcegraph/issues/55429: **1d**
- [ ] https://github.com/sourcegraph/sourcegraph/issues/55430: **2d**
- [ ] https://github.com/sourcegraph/sourcegraph/issues/55428: **1d**

#### ❗ Risks

- The estimated work is for 5 wall clock days and I now have around 4.5 days left
- Some work admin stuff with deadlines this week:
  - I need to write 4 impact reviews
  - I need to complete the `Continu` training
- I am going on PTO at the end of this week

All things considered, this is a bit too tight and close for comfort and my one line goal for this week is to complete any open action items on this project and leave with docs and planned project board so that if anyone wants to pick it up while I'm gone, it is easy for them to slot it. 🤞
